lib/signaling/v3/room.js

```diff
@@ -0,0 +1,158 @@
+      participantState,
+      trackSid => this._getPendingTrackReceiver(trackSid),
+      trackSid => this._getInitialTrackSwitchOffState(trackSid),
+      (trackSid, priority) => this._trackPrioritySignaling.sendTrackPriorityUpdate(trackSid, 'subscribe', priority),
+      (trackSid, hint) => this._renderHintsSignaling.setTrackHint(trackSid, hint),
+      trackSid => this._renderHintsSignaling.clearTrackHint(trackSid)
+    );
+  }
+
+  /**
+   * @private
+   * @override
+   */
+  _getInitialTrackSwitchOffState(trackSid) {
+    const switchOffState = this._pendingSwitchOffStates.get(trackSid)
+      || { state: 'OFF', switchOffReason: 'DISABLED_BY_SUBSCRIBER' };
+    this._pendingSwitchOffStates.delete(trackSid);
+    if (switchOffState.state === 'OFF') {
+      this._log.warn(`[${trackSid}] was initially switched off! `);
+    }
+    return switchOffState;
+  }
+
+  /**
+   * @private
+   */
+  _getPendingTrackReceiver(trackSid) {
+    const mid = this._pendingTrackMids.get(trackSid);
+    if (!mid) {
+      return Promise.resolve(null);
+    }
+    this._pendingTrackMids.delete(trackSid);
+    return this._getTrackReceiver(mid, 'mid');
+  }
+
+  /**
+   * @private
+   * @override
+   */
+  _handleSubscriptions() {
+    /* Do nothing since RSP v3 messages will not contain the "subscribed" property. */
+  }
+
+  /**
+   * @private
+   */
+  _initTrackSubscriptionsSignaling() {
+    this._trackSubscriptionsSignaling.on('updated', (subscribed, errors) => {
+      const trackSidsToTrackSignalings = this._getTrackSidsToTrackSignalings();
+      const trackSidsToTrackStates = new Map(Object.entries(subscribed));
+      const trackSidsToErrors = new Map(Object.entries(errors));
+
+      trackSidsToTrackStates.forEach(({ mid, off_reason: switchOffReason = null, state }, sid) => {
+        const trackSignaling = trackSidsToTrackSignalings.get(sid);
+        const trackState = { state, switchOffReason };
+        if (!trackSignaling) {
+          this._pendingSwitchOffStates.set(sid, trackState);
+          this._pendingTrackMids.set(sid, mid);
+          return;
+        }
+        const isSwitchedOff = state === 'OFF';
+        if (isSwitchedOff || (trackSignaling.trackTransceiver && trackSignaling.trackTransceiver.mid !== mid)) {
+          // NOTE(mmalavalli): If a RemoteTrackPublicationV3's MID changes, then we need to unsubscribe
+          // from the RemoteTrack before subscribing to it again with the MediaTrackReceiver associated with the new
+          // MID. If a RemoteTrackPublicationV3's RemoteTrack is switched off, then we should still be subscribed
+          // to it, even though it no longer has an MID associated with it.
+          trackSignaling.setTrackTransceiver(null, isSwitchedOff);
+        }
+        if (!isSwitchedOff) {
+          this._getTrackReceiver(mid, 'mid').then(trackReceiver => trackSignaling.setTrackTransceiver(trackReceiver, true));
+        }
+        trackSignaling.setSwitchedOff(isSwitchedOff);
```

Comment:
I think you are coupling the `state` and transceiver management together slightly by disallowing an `'OFF'` state with an assigned MID. I know that the server should not do this, and the protocol spec says it should not be done but I don't know if the client should also enforce the same restriction at the room level without even raising warnings to the developer.

Reply:
@ceaglest I think the code is written with the server behavior as described in the blueprint /  MSP spec. What you are saying is only useful if the SFU introduces a bug into prof where an MID is assigned to a switched off track. In my view the SFU code should have a test where it makes sure that this doesn't happen.